packages/gatsby/src/schema/node-model.js

```diff
@@ -51,14 +56,26 @@ export interface NodeModel {
     result: nodeOrNodes,
     pageDependencies?: PageDependencies
   ): nodesOrNodes;
+  findRootNodeAncestor(obj: any, predicate: () => boolean): Node | null;
+  trackInlineObjectsInRootNode(node: Node, sanitize: boolean): Node;
 }
 
 class LocalNodeModel {
-  constructor({ schema, nodeStore, createPageDependency, path }) {
+  constructor({ schema, schemaComposer, nodeStore, createPageDependency }) {
     this.schema = schema
+    this.schemaComposer = schemaComposer
     this.nodeStore = nodeStore
     this.createPageDependency = createPageDependency
-    this.path = path
+
+    this._rootNodeMap = new WeakMap()
+    this._trackedRootNodes = new Set()
+    this._prepareNodesQueues = {}
+    this._prepareNodesPromises = {}
+    this._preparedNodesCache = new Map()
+  }
+
+  withContext(context) {
+    return new ContextualNodeModel(this, context)
```

Comment:
I do wonder why we not augment this model instead?

Reply:
There can be multiple contextual node models rooting from one node model. They share state, but have different paths.